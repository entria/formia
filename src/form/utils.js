import get from 'lodash.get';
import set from 'lodash.set';
import isEqual from 'lodash.isequal';

export const handleInitialValuesUpdate = (
  currentValues = {},
  currentInitialValues = {},
  newInitialValues = {},
) => {
  const updatedValues = { ...currentValues };

  Object.keys(newInitialValues).map(fieldName => {
    const fieldCurrentValue = get(currentValues, fieldName, null);
    const fieldCurrentInitialValue = get(currentInitialValues, fieldName, null);
    const isFieldPristine = isEqual(fieldCurrentValue, fieldCurrentInitialValue);

    if (!isFieldPristine) {
      return;
    }

    const fieldNewInitialValue = get(newInitialValues, fieldName, null);
    set(updatedValues, fieldName, fieldNewInitialValue);
  });

  return updatedValues;
};
