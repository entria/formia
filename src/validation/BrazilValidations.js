// @flow
import { Strings } from '@entria/utils';

import type { RuleError } from './Validation';

export const cpf = () => {
  const validateFirstDigit = (value: any): boolean => {
    let Soma = 0;
    for (let i = 1; i <= 9; i++) {
      Soma = Soma + parseInt(value.substring(i - 1, i), 10) * (11 - i);
    }

    let Resto = Soma * 10 % 11;
    if (Resto === 10 || Resto === 11) {
      Resto = 0;
    }

    return Resto === parseInt(value.substring(9, 10), 10);
  };

  const validateSecondDigit = (value: any): boolean => {
    let Soma = 0;
    for (let i = 1; i <= 10; i++) {
      Soma = Soma + parseInt(value.substring(i - 1, i), 10) * (12 - i);
    }

    let Resto = Soma * 10 % 11;
    if (Resto === 10 || Resto === 11) {
      Resto = 0;
    }

    return Resto === parseInt(value.substring(10, 11), 10);
  };

  return (value: any): ?RuleError => {
    if (Strings.isEmpty(value)) {
      return null;
    }

    const sanitizedValue = Strings.numbers(value);
    if (sanitizedValue === '00000000000') {
      return {
        code: 'Validation.Brazil.cpf',
        message: 'Invalid CPF',
        params: [],
      };
    }

    const isFirstDigitValid = validateFirstDigit(sanitizedValue);
    const isSecondDigitValid = validateSecondDigit(sanitizedValue);
    const isValid = isFirstDigitValid && isSecondDigitValid;
    if (isValid) {
      return null;
    }

    return {
      code: 'Validation.Brazil.cpf',
      message: 'Invalid CPF',
      params: [],
    };
  };
};

export const phone = () => (value: any): ?RuleError => {
  if (Strings.isEmpty(value)) {
    return null;
  }

  const sanitizedValue = Strings.numbers(value);
  const thirdNumber = sanitizedValue.substr(2, 1);

  const validCellphone = sanitizedValue.length === 11 ? thirdNumber === '9' : true;
  const validLandline = sanitizedValue.length === 10 ? thirdNumber !== '0' : true;
  const isValid = validCellphone && validLandline;
  if (isValid) {
    return null;
  }

  return {
    code: 'Validation.Brazil.phone',
    message: 'Invalid phone number',
    params: [],
  };
};
