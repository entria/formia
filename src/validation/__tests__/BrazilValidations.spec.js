import * as Brazil from '../BrazilValidations';

import { setupTest } from '../../../test/utils';

beforeEach(async () => setupTest());

it('cpf - empty', () => {
  expect(Brazil.cpf()()).toMatchSnapshot();
});

it('cpf - valid document', () => {
  expect(Brazil.cpf()('15516405114')).toMatchSnapshot();
});

it('cpf - bunch of zeros', () => {
  expect(Brazil.cpf()('00000000000')).toMatchSnapshot();
});

it('cpf - invalid', () => {
  expect(Brazil.cpf()('12345678900')).toMatchSnapshot();
});

it('phone - empty', () => {
  expect(Brazil.phone()()).toMatchSnapshot();
});

it('phone - valid cellphone', () => {
  expect(Brazil.phone()('49991237830')).toMatchSnapshot();
});

it('phone - valid landline', () => {
  expect(Brazil.phone()('4936310000')).toMatchSnapshot();
});

it('phone - invalid cellphone', () => {
  expect(Brazil.phone()('49012345678')).toMatchSnapshot();
});

it('phone - invalid landline', () => {
  expect(Brazil.phone()('4901234566')).toMatchSnapshot();
});
