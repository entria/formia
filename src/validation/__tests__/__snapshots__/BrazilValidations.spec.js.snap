// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cpf - bunch of zeros 1`] = `
Object {
  "code": "Validation.Brazil.cpf",
  "message": "Invalid CPF",
}
`;

exports[`cpf - empty 1`] = `null`;

exports[`cpf - invalid 1`] = `
Object {
  "code": "Validation.Brazil.cpf",
  "message": "Invalid CPF",
}
`;

exports[`cpf - valid document 1`] = `null`;

exports[`phone - empty 1`] = `null`;

exports[`phone - invalid cellphone 1`] = `
Object {
  "code": "Validation.Brazil.phone",
  "message": "Invalid phone number",
}
`;

exports[`phone - invalid landline 1`] = `
Object {
  "code": "Validation.Brazil.phone",
  "message": "Invalid phone number",
}
`;

exports[`phone - valid cellphone 1`] = `null`;

exports[`phone - valid landline 1`] = `null`;
